### ISSUE TITLE
potential performance bottleneck

### ISSUE THREAD

**Issue by bpatelcs** [Yelp/py_zipkin#29]
@mjbryant  I am new to advanced python, feel free to scold me if I am wrong.

so when we want to trace the function call we can use decorator but by looking at the following code it looks like for each decorator declaration it will create two zipkin instances and throw the first one.

when we do `@zipkin(service_name="foo",span_name="bar")` it will first create the instance of zipkin class by passing the arguments to `__init__` and then the `__call__` method will be called and again in the decorated function you are creating the new instance with the same arguments in context manager block , in my view we should optimize the code if this is true.

    def __call__(self, f):
        @functools.wraps(f)
        def decorated(*args, **kwargs):
            with zipkin_span(
                service_name=self.service_name,
                span_name=self.span_name,
                zipkin_attrs=self.zipkin_attrs,
                transport_handler=self.transport_handler,
                annotations=self.annotations,
                binary_annotations=self.binary_annotations,
                port=self.port,
                sample_rate=self.sample_rate,
                include=self.include,
            ):
                return f(*args, **kwargs)
        return decorated

**Comment by mjbryant** [comment]
@bpatelcs first - thanks for the issue!

I'm not entirely sure what you're suggesting. This particular form of the `__call__` function was adopted to fix #20. The root cause of this issue is the fact that decorators on class methods or globally scoped functions are evaluated at import time, not at runtime. So attributes passed into the `zipkin_span`'s `__init__` method might have incorrect values.

See https://github.com/Yelp/py_zipkin/pull/21 for the fix.

**Comment by kaisen** [comment]
@bpatelcs good observation, but the issue that @mjbryant linked is the reason why the code is structured this way. Ultimately, the extra __init__ and __call__ calls will only happen once, so the performance hit is negligible.

```python
import functools


class context_manager_and_decorator(object):

    def __init__(self):
        print 'in __init__'

    def __call__(self, f):
        print 'in __call__'
        @functools.wraps(f)
        def decorated(*args, **kwargs):
            with context_manager_and_decorator():
                return f(*args, **kwargs)
        return decorated

    def __enter__(self):
        print 'in __enter__'
        return self

    def __exit__(self, _exc_type, _exc_value, _exc_traceback):
        print 'in __exit__'


@context_manager_and_decorator()
def decorated_function():
    print 'in decorated_function'

print ''

decorated_function()

print ''

with context_manager_and_decorator():
    print 'in context'
```
